CoreIPC should be explicit about which methods can be called when WaitingForSyncReply

gavin-apple opened the pull request:
#### 58b5d6c6019cbd67896c50af8ae6b6ea94165f1d
<pre>
CoreIPC should be explicit about which methods can be called when WaitingForSyncReply

<a href="https://bugs.webkit.org/show_bug.cgi?id=247673">https://bugs.webkit.org/show_bug.cgi?id=247673</a>
rdar://102134965

Reviewed by Alex Christensen.

CoreIPC should be explicit about which methods can be called when
WaitingForSyncReply. This change adds a new message attribute,
AllowedWhenWaitingForSyncReply, which specifies a message should be allowed.
Any messages which don&apos;t have this attribute will be queued as normal.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::dispatchSyncMessage):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::isAllowedWhenWaitingForSyncReply const):
(IPC::Decoder::isAllowedWhenWaitingForUnboundedSyncReply const):
(IPC::Decoder::setIsAllowedWhenWaitingForSyncReplyOverride):
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Scripts/webkit/model.py:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::Subnamespace::StructName&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::ReturnRefClass&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritsFrom&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritanceGrandchild&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::messageAllowedWhenWaitingForSyncReply):
(IPC::messageAllowedWhenWaitingForSyncReplyDueToSynchronousAttribute):
(IPC::messageAllowedWhenWaitingForSyncReplyDueToStreamAttribute):
(IPC::messageAllowedWhenWaitingForUnboundedSyncReply):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/257229@main">https://commits.webkit.org/257229@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/abdfcb0a888dd8aee22e932f9b9e15f3639f6011

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31272 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107581 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167847 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7824 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36098 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90724 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104186 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5883 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84716 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33010 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/101667 "webkitpy-tests (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87753 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89466 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1305 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20898 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1266 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22401 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4981 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6131 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44854 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2574 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41814 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -82,9 +82,9 @@ Decoder::Decoder(const uint8_t* buffer, size_t bufferSize, BufferDeallocator&& b
+    
     if (UNLIKELY(!decode(m_messageFlags)))
-        return;
+            return;
```
Incorrectly added indentation.

achristensen07 commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
```diff
@@ -22,7 +22,7 @@
 #if ENABLE(GPU_PROCESS)
 
-messages -> RemoteDisplayListRecorder NotRefCounted Stream {
+messages -> RemoteDisplayListRecorder NotRefCounted Stream AllowedWhenWaitingForSyncReply {
```
I think in WebKit currently we can make Stream imply AllowedWhenWaitingForSyncReply

gavin-apple replied:
ack. That's now the case - I must of missed this one. 

Currently Stream ==> AllowedWhenWaitingForSyncReply (for now) but AllowedWhenWaitingForSyncReply !==> Stream

achristensen07 commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -82,7 +82,7 @@ Decoder::Decoder(const uint8_t* buffer, size_t bufferSize, BufferDeallocator&& b
         return;
     }
-
+    
```
whitespace

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/MessageNames.h:
```diff
@@ -145,6 +145,32 @@ constexpr bool messageIsSync(MessageName name)
     return name >= MessageName::TestWithLegacyReceiver_GetPluginProcessConnection;
 }
 
+constexpr bool messageAllowedWhenWaitingForSyncReply(MessageName name)
```
Let's not put these implementations in the header.  It'll hurt compile times, unnecessarily increase binary size, and probably not improve performance significantly.

achristensen07 commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1068,6 +1070,7 @@ void Connection::dispatchSyncMessage(Decoder& decoder)
         }
         std::unique_ptr<Decoder> unwrappedDecoder = Decoder::unwrapForTesting(decoder);
         RELEASE_ASSERT(unwrappedDecoder);
+        unwrappedDecoder->setIsAllowedWhenWaitingForSyncReplyOverride(true);
```
This seems redundant with messageAllowedWhenWaitingForSyncReplyDueToSynchronousAttribute

gavin-apple replied:
If you take a look at void Encoder::wrapForTesting(UniqueRef<Encoder>&& original), you'll see it's another spot where we dynamically add ShouldDispatchWhenWaitingForSyncReply::Yes to all messages when we use m_inDispatchMessageMarkedToUseFullySynchronousModeForTesting. So this change mirrors the expansion of the allow list on the receiving side when in that mode of testing.

I'll move it into the corresponding unwrapForTesting - which makes more sense....

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -115,11 +115,11 @@ std::optional<Namespace::Subnamespace::StructName> ArgumentCoder<Namespace::Subn
     return {
         Namespace::Subnamespace::StructName {
-            WTFMove(*firstMemberName)
+            WTFMove(*firstMemberName),
```
This is probably not a part of this PR.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/MessageNames.h:
```diff
@@ -145,6 +145,10 @@ constexpr bool messageIsSync(MessageName name)
 
+bool messageAllowedWhenWaitingForSyncReply(MessageName);
+bool messageAllowedWhenWaitingForUnboundedSyncReply(MessageName);
+bool messageAllowedWhenWaitingForSyncReplyDueToStreamAttribute(MessageName);
```
I would actually remove this function and make only messageAllowedWhenWaitingForSyncReply and messageAllowedWhenWaitingForUnboundedSyncReply and have their implementations return true for multiple reasons.  That will reduce the number of places where we have the opportunity to forget to check two functions.

Or we could make messageAllowedWhenWaitingForSyncReply be the only caller of messageAllowedWhenWaitingForSyncReplyDueToStreamAttribute and messageAllowedWhenWaitingForSyncReplyDueToSynchronousAttribute which would then be static functions not declared in the header.

gavin-apple replied:
I've changed the logic to let messageAllowedWhenWaitingForSyncReply return true for multiple reasons.

gavin-apple committed:
CoreIPC should be explicit about which methods can be called when WaitingForSyncReply

https://bugs.webkit.org/show_bug.cgi?id=247673
rdar://102134965

Reviewed by Alex Christensen.

CoreIPC should be explicit about which methods can be called when
WaitingForSyncReply. This change adds a new message attribute,
AllowedWhenWaitingForSyncReply, which specifies a message should be allowed.
Any messages which don't have this attribute will be queued as normal.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::dispatchSyncMessage):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::isAllowedWhenWaitingForSyncReply const):
(IPC::Decoder::isAllowedWhenWaitingForUnboundedSyncReply const):
(IPC::Decoder::setIsAllowedWhenWaitingForSyncReplyOverride):
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
(generate_message_names_implementation):
* Source/WebKit/Scripts/webkit/model.py:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::Subnamespace::StructName>::decode):
(IPC::ArgumentCoder<Namespace::OtherClass>::decode):
(IPC::ArgumentCoder<Namespace::ReturnRefClass>::decode):
(IPC::ArgumentCoder<WebCore::InheritsFrom>::decode):
(IPC::ArgumentCoder<WebCore::InheritanceGrandchild>::decode):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::messageAllowedWhenWaitingForSyncReply):
(IPC::messageAllowedWhenWaitingForSyncReplyDueToSynchronousAttribute):
(IPC::messageAllowedWhenWaitingForSyncReplyDueToStreamAttribute):
(IPC::messageAllowedWhenWaitingForUnboundedSyncReply):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/257229@main
